torch/csrc/jit/passes/inline_forked_closures.cpp

```diff
@@ -58,6 +58,88 @@ void inlineForkedClosure(Node* fork_closure, NodeKind genKind) {
   runCleanupPasses(fork_graph);
 }
 
+void inlineAwaitableThenClosure(Node* then_closure) {
+  Node* function_context_node = then_closure->inputs()[0]->node();
+
+  if (function_context_node->inputs().size() != 2 ||
+      function_context_node->inputs().at(0)->node()->kind() != prim::Closure ||
+      function_context_node->inputs().at(1)->node()->kind() !=
+          prim::TupleConstruct) {
+    throw ErrorReport(then_closure->sourceRange()) << "Cannot fork this value";
+  }
+
+  Node* function = function_context_node->inputs().at(0)->node();
+  Node* context = function_context_node->inputs().at(1)->node();
+  auto then_graph = function->g(attr::Subgraph)->copy();
+  // Replacing ir_emitter emitted awaitableThenClosure, with awaitable_then node
+  // with 2 inputs: Await Value which graph it continues, the fake input Value
+  // which represents the future result of Await graph with previously added
+  // continuations execution. As we do not have Value for it in IR - we add
+  // no-op node awaitable_then_input to produce this Value and keep the graph
+  // valid. At runtime instead the result of Await will be placed for this
```

Comment:
Thanks a lot for this.

nit: I suggested how to re-phrase this below. (I find this confusing so I think it would be good to be very clear in the comments, so future developers have a better chance of understanding!). Please make sure what I wrote makes sense (and add the part I was wondering about)

> ir_emitter produces IR with prim::awaitableThenClosure; but just like the inlineForkedClosure method above, we need to modify it before running it.
> We replace the prim::Closure with a prim::awaitable_then, which is a callable.
> Note that a prim::awaitable_then function takes arguments (%x: Await(T), %y: T). In the calling graph, we will have %x: Await(T) available; it's the awaitable that the ops are operating on. However, we won't have second argument %y: T. (At execution time, %y, the input to the function we're calling as part of our awaitable_then call, will be the output of the previous function queued in the Awaitable object). As a workaround, we use prim::awaitable_then_input to create a dummy argument in the calling graph that we pass into the awaitable_then, which makes the graph valid.

...
TODO - @IvanKobzarev : can you comment about needing to unpack the context tuple? Will the context tuple ever be non-empty?
...

Reply:
Thanks a lot for the comment wording, this is historically one of the hardest parts for me :) Changed the comment.